src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -2458,12 +2513,43 @@ Class<?> defineClass(boolean initialize, Object classData) {
                 Class<?> lookupClass = lookup.lookupClass();
                 ClassLoader loader = lookupClass.getClassLoader();
                 ProtectionDomain pd = (loader != null) ? lookup.lookupClassProtectionDomain() : null;
-                Class<?> c = SharedSecrets.getJavaLangAccess()
-                        .defineClass(loader, lookupClass, name, bytes, pd, initialize, classFlags, classData);
-                assert !isNestmate() || c.getNestHost() == lookupClass.getNestHost();
-                return c;
+                Class<?> c = null;
+                try {
+                    c = SharedSecrets.getJavaLangAccess()
+                            .defineClass(loader, lookupClass, name, bytes, pd, initialize, classFlags, classData);
+                    assert !isNestmate() || c.getNestHost() == lookupClass.getNestHost();
+                    return c;
+                } finally {
+                    // dump the classfile for debugging
+                    if (dumper.isEnabled()) {
+                        String name = internalName();
+                        if (c != null) {
+                            String cn = c.getName();
+                            int suffixIdx = cn.lastIndexOf('/');
+                            if (suffixIdx > 0) {
+                                assert ((classFlags & HIDDEN_CLASS) != 0);
+                                name += '.' + cn.substring(suffixIdx + 1);
+                            } else {
+                                assert ((classFlags & HIDDEN_CLASS) == 0);
+                            }
+                        } else {
+                            name += ".failed-" + dumper.incrementAndGetCounter();
```

Comment:
I think it makes more sense to move the counter to `ClassDefiner`. It is not used by `ClassFileDumper` itself.

(make it `static` if it's possible to have multiple definer instances with the same class name)

Reply:
I see your point.   If it were a static counter for all dumpers,  multiple`.failed-xxx` dumped by a single dumper may not be in sequence if other dumpers have `.failed-xxx` class files.

Reply:
If the counter has to be per dumper, maybe it makes sense to push the logic that derives the file name into ClassFileDumper too. e.g. have a `dumpClass(name, Class<?>, bytes)` and `dumpFailed(name, bytes)`.

Reply:
that's a good idea.  Updated.